### PR TITLE
Updated the usage of TensorMap in generated kernels

### PR DESCRIPTION
**Pull request by drzejan2**
The scope of introduced changes:
- modify the string representation of TensorMap object type for kernel
  lowering phase,
- fix the type of TensorMap local variables in kernel (aliases),
- add local helper tool to detect TensorMap data type,

### CODE REVIEW

**Comment by drzejan2** [comment]
I checked the following test cases with the proposed changes on h100 device:
```
TMATest.LoadCompleteTensor1D
TMATest.LoadCompleteTensor2D
TMATest.LoadCompleteTensor3D
TMATest.LoadCompleteTensor4D
TMATest.LoadCompleteTensor5D
TMATest.StoreCompleteTensor1D
TMATest.StoreCompleteTensor2D
TMATest.StoreCompleteTensor3D
TMATest.StoreCompleteTensor4D
TMATest.StoreCompleteTensor5D
TMATest.DisableIndexHoisting
```
there were no issues, all tests passed.

---

The issue surfaced during the work on Double Buffering / Circular buffer support for Hopper, where the kernel initially looked like this:

```
__global__ void nvfuser_none_f0_c0_r0_g0(Tensor<float, 1, 1> T0, const __grid_constant__ TensorMap var0, Tensor<float, 1, 1> T2) {
  alignas(16) extern __shared__ char array[];
  const unsigned smem_offset = 0;
  NVFUSER_DEFINE_MAGIC_ZERO;
  nvfuser_index_t i1;
  i1 = ceilDiv(T0.logical_size[0LL], 32);
  const __grid_constant__ TensorMap* ptr2;
  ptr2 = &var0;
...
...
CUDA NVRTC compile error: __tmp_kernel_none_f0_c0_r0_g0.cu(9451): error: attribute "grid_constant" does not apply here
    const __grid_constant__ TensorMap* ptr2;
          ^
```

With the change (the kernel still does not compile, but due to other issues):
```
__global__ void nvfuser_none_f0_c0_r0_g0(Tensor<float, 1, 1> T0, const __grid_constant__ TensorMap var0, Tensor<float, 1, 1> T2) {
  alignas(16) extern __shared__ char array[];
  const unsigned smem_offset = 0;
  NVFUSER_DEFINE_MAGIC_ZERO;
  nvfuser_index_t i1;
  i1 = ceilDiv(T0.logical_size[0LL], 32);
  const TensorMap* ptr2;
  ptr2 = &var0;
...
```

**Comment by drzejan2** [comment]
!build

File: csrc/codegen.cpp [hunk 1]
```diff
@@ -34,6 +34,25 @@ std::string ptrType(DataType dt) {
...
+        } else if constexpr (std::is_same_v<T, OpaqueType>) {
+          if (0 == dtype.name.compare("TensorMap")) {
+            return true;
+          } else {
+            return false;
+          }
+        } else {
+          return false;
+        }
+      },
```
Comment (wujingyue) [inline csrc/codegen.cpp:52]:
```suggestion
      [](auto&& dtype) -> bool {
        using T = std::decay_t<decltype(dtype)>;
        if constexpr (std::is_same_v<T, PointerType>) {
          return isTmaType(*dtype.type);
        }
        if constexpr (std::is_same_v<T, OpaqueType>) {
          return 0 == dtype.name.compare("TensorMap");
        }
        return false;
      },
```

**Review by wujingyue — COMMENTED** [review]
A drive-by comment

**Review by zasdfgbnm — APPROVED** [review]
Jingyue's suggestion does make sense to me. Other than that, LGTM!

File: csrc/codegen.cpp (hunk 1 above)
Comment (drzejan2) [reply csrc/codegen.cpp:52]: Thanks for suggestion, @wujingyue - you are right, this should be simplified. I will update the code. Thanks!

I made it like this as I wanted to keep it aligned with how `data_type2string(...)` is implemented, but that's an overkill.

**Comment by drzejan2** [comment]
!build

**Pull request merged**